File: monitoring-backend/app/schemas/alert.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class AlertResponse(BaseModel):
    alert_id: int
    device_id: Optional[int] = None
    switch_id: Optional[int] = None
    device_name: Optional[str] = None
    location_name: Optional[str] = None
    librenms_alert_id: Optional[int]
    category_id: Optional[int]
    alert_type: str
    severity: str
    message: str
    status: str
    assigned_to_user_id: Optional[int] = None
    resolved_by_full_name: Optional[str] = None
    acknowledged_at: Optional[datetime] = None
    resolution_note: Optional[str] = None
    created_at: datetime
    cleared_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AlertUpdate(BaseModel):
    resolution_note: Optional[str] = Field(None, max_length=2000)
    assigned_to_user_id: Optional[int] = None


class AlertFilters(BaseModel):
    status: Optional[str] = None
    severity: Optional[str] = None
    alert_type: Optional[str] = None
    device_id: Optional[int] = None
    switch_id: Optional[int] = None
    category_id: Optional[int] = None


class AlertPageResponse(BaseModel):
    items: List[AlertResponse]
    total: int
    page: int
    page_size: int


class AlertBulkDeleteResponse(BaseModel):
    deleted: int

File: monitoring-frontend/lib/models/alert.dart
class Alert {
  final int alertId;
  final int? deviceId;
  final int? switchId;
  final String deviceName;
  final String locationName;
  final String alertType;
  final String severity;
  final String message;
  final String status;
  final int? assignedToUserId; //this act as acknowledged by for now
  final String? resolvedByFullName;
  final DateTime createdAt;
  final DateTime? clearedAt;
  final DateTime? acknowledgedAt;
  final String? resolutionNote;

  Alert({
    required this.alertId,
    this.deviceId,
    this.switchId,
    required this.deviceName,
    required this.locationName,
    required this.alertType,
    required this.severity,
    required this.message,
    required this.status,
    this.assignedToUserId,
    this.resolvedByFullName,
    required this.createdAt,
    this.clearedAt,
    this.acknowledgedAt,
    this.resolutionNote,
  });

  factory Alert.fromJson(Map<String, dynamic> json) {
    return Alert(
      alertId: json['alert_id'],
      deviceId: json['device_id'],
      switchId: json['switch_id'],
      deviceName: json['device_name'] ?? ' - ',
      locationName: json['location_name'] ?? ' - ',
      alertType: json['alert_type'] ?? 'Unknown',
      severity: json['severity'] ?? 'info',
      message: json['message'] ?? '',
      status: json['status'] ?? 'active',
      assignedToUserId: json['assigned_to_user_id'],
      resolvedByFullName: json['resolved_by_full_name'],
      createdAt: DateTime.parse(json['created_at']),
      clearedAt: json['cleared_at'] != null
          ? DateTime.parse(json['cleared_at'])
          : null,
      acknowledgedAt: json['acknowledged_at'] != null
          ? DateTime.parse(json['acknowledged_at'])
          : null,
      resolutionNote: json['resolution_note'],
    );
  }
}
